csrc/utils.h

```diff
@@ -386,28 +388,70 @@ std::string toDelimitedInlineString(
   return ss.str();
 }
 
-template <typename... Args>
 class DebugPrintScope {
  public:
+  template <typename... Args>
   DebugPrintScope(std::string name, Args... args) : name_(std::move(name)) {
     debug() << "Entering " << name_ << "("
```

Comment:
When we say its `debug()`, is it actually disabled when `NDEBUG` or just the same as std::cerr?

Reply:
I think it refers to the `debug` function in `debug.h`, which is just a utility for capturing the output printing. I think at some point we received a request from user about capturing the debug output, and this is how we help with this.

Reply:
Yes, this lets us send the output to python instead of stdout, if necessary. In the future it might be nice to have structured logging where we could use multiple streams, but for now, there is a guard that can be used to redirect all `debug()` output to a custom `ostream`. It defaults to stdout, not stderr.

Reply:
Yes, this lets us send the output to python instead of stdout, if necessary. In the future it might be nice to have structured logging where we could use multiple streams, but for now, there is a guard that can be used to redirect all `debug()` output to a custom ostream. It defaults to stdout, not stderr.